File: 프로그래머스/3/43238. 입국심사/입국심사.py
import sys
def solution(n, times):
    answer = 0
    start = 1
    end = sys.maxsize
    
    while start <= end:
        mid = (start + end) // 2
        
        cnt = 0
        for time in times:
            cnt += mid // time
        
        if cnt >= n:
            end = mid - 1
        else:
            start = mid + 1

    return start

File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;
import java.io.*;

class Solution {
    public long solution(int n, int[] times) {

        long start = 1;
        long end =  (long)times[times.length-1] * n;
        
        while (start <= end) {
            long mid = (start + end) / 2;
            
            long total = 0;
            
            for (int time : times) {
                total += mid / time;
            }
            
            if (total >= n) end = mid - 1;
            else start = mid + 1;
        }
        
        return start;
    }
}